File: app/src/main/java/com/vmloft/develop/app/match/ui/sign/SignUpActivity.java
package com.vmloft.develop.app.match.ui.sign;

import android.app.Dialog;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;

import com.vmloft.develop.app.match.R;
import com.vmloft.develop.app.match.base.ACallback;
import com.vmloft.develop.app.match.base.AppActivity;
import com.vmloft.develop.app.match.request.bean.AAccount;
import com.vmloft.develop.app.match.common.ASignManager;
import com.vmloft.develop.library.tools.utils.VMReg;
import com.vmloft.develop.library.tools.utils.VMStr;

import butterknife.BindView;
import butterknife.OnClick;

import com.vmloft.develop.library.tools.utils.VMTheme;
import com.vmloft.develop.library.tools.widget.VMEditView;
import com.vmloft.develop.library.tools.widget.toast.VMToast;

/**
 * Create by lzan13 on 2019/05/09
 *
 * 注册界面
 */
public class SignUpActivity extends AppActivity {

    // 输入框
    @BindView(R.id.sign_account_et) VMEditView mAccountView;
    @BindView(R.id.sign_password_et) VMEditView mPasswordView;
    @BindView(R.id.sign_up_btn) Button mSignUpBtn;

    private String mAccount;
    private String mPassword;

    private Dialog mDialog;

    @Override
    protected int layoutId() {
        return R.layout.activity_sign_up;
    }

    @Override
    protected void initUI() {
        super.initUI();

        VMTheme.changeShadow(mAccountView);
        VMTheme.changeShadow(mPasswordView);

        // 监听输入框变化
        mAccountView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                verifyInputBox();
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
        mPasswordView.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                verifyInputBox();
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });
    }

    @Override
    protected void initData() {

    }

    /**
     * 界面内控件的点击事件监听器
     */
    @OnClick({ R.id.sign_up_btn })
    void onClick(View v) {
        switch (v.getId()) {
        case R.id.sign_up_btn:
            registerByEmail();
            break;
        }
    }

    /**
     * 校验输入框内容
     */
    private void verifyInputBox() {
        // 将用户名转为消息并修剪
        mAccount = mAccountView.getText();
        mPassword = mPasswordView.getText();

        // 检查输入框是否为空是否为空
        if (VMStr.isEmpty(mPassword) || VMStr.isEmpty(mAccount)) {
            mSignUpBtn.setEnabled(false);
        } else {
            mSignUpBtn.setEnabled(true);
        }
    }

    /**
     * 通过邮箱注册
     */
    private void registerByEmail() {
        if (!VMReg.isEmail(mAccount)) {
            VMToast.make(mActivity, VMStr.byRes(R.string.account_is_email)).error();
            return;
        }
        showDialog();
        ASignManager.getInstance().signUpByEmail(mAccount, mPassword, new ACallback<AAccount>() {
            @Override
            public void onSuccess(AAccount account) {
                if (mDialog != null) {
                    mDialog.dismiss();
                }
                VMToast.make(mActivity, R.string.sign_up_success).show();
                // 注册成功保存下用户信息，方便回到登录页面输入信息
                ASignManager.getInstance().setHistoryAccount(account);
                onFinish();
            }

            @Override
            public void onError(int code, String desc) {
                if (mDialog != null) {
                    mDialog.dismiss();
                }
                VMToast.make(mActivity, desc).error();
            }
        });
    }

    /**
     * 显示对话框
     */
    private void showDialog() {
        mDialog = new Dialog(mActivity);
        mDialog.setContentView(R.layout.widget_custom_progress_dialog);
        mDialog.show();
    }

    @Override
    protected void onDestroy() {
        if (mDialog != null) {
            mDialog.dismiss();
        }
        super.onDestroy();
    }
}

File: app/src/main/java/com/vmloft/develop/app/template/ui/sign/SignUpActivity.kt
package com.vmloft.develop.app.template.ui.sign


import android.text.Editable
import android.text.TextWatcher
import android.text.method.HideReturnsTransformationMethod
import android.text.method.PasswordTransformationMethod
import android.view.View

import com.alibaba.android.arouter.facade.annotation.Route

import com.vmloft.develop.app.template.R
import com.vmloft.develop.app.template.databinding.ActivitySignUpBinding
import com.vmloft.develop.app.template.request.viewmodel.SignViewModel
import com.vmloft.develop.app.template.common.Constants
import com.vmloft.develop.app.template.im.IMManager
import com.vmloft.develop.app.template.router.AppRouter
import com.vmloft.develop.library.base.BVMActivity
import com.vmloft.develop.library.base.BViewModel
import com.vmloft.develop.library.base.event.LDEventBus
import com.vmloft.develop.library.base.router.CRouter
import com.vmloft.develop.library.base.utils.errorBar
import com.vmloft.develop.library.data.common.SignManager
import com.vmloft.develop.library.tools.utils.VMReg

import org.koin.androidx.viewmodel.ext.android.getViewModel


/**
 * Create by lzan13 on 2020/6/19 17:10
 * 描述：注册界面
 */
@Route(path = AppRouter.appSignUp)
class SignUpActivity : BVMActivity<ActivitySignUpBinding, SignViewModel>() {

    private var mAccount: String = ""
    private var mPassword: String = ""

    override fun initVM(): SignViewModel = getViewModel()

    override fun initVB() = ActivitySignUpBinding.inflate(layoutInflater)

    override fun initUI() {
        super.initUI()
        setTopTitle(R.string.sign_up)

        // 监听输入框变化
        mBinding.accountET.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                mAccount = s.toString().trim()
                verifyInputBox()
            }
        })
        mBinding.passwordET.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable) {
                mPassword = s.toString().trim()
                verifyInputBox()
            }
        })

        // 设置密码隐藏与显示
        mBinding.passwordIconIV.setOnClickListener {
            if (mBinding.passwordIconIV.isSelected) {
                // 隐藏密码
                mBinding.passwordET.transformationMethod = PasswordTransformationMethod.getInstance()
                mBinding.passwordIconIV.isSelected = false
            } else {
                // 显示密码
                mBinding.passwordET.transformationMethod = HideReturnsTransformationMethod.getInstance()
                mBinding.passwordIconIV.isSelected = true
            }
        }
        mBinding.userAgreementTV.setOnClickListener { CRouter.go(AppRouter.appSettingsAgreementPolicy, str0 = "agreement") }
        mBinding.privatePolicyTV.setOnClickListener { CRouter.go(AppRouter.appSettingsAgreementPolicy, str0 = "policy") }
        mBinding.submitTV.setOnClickListener {
            if (mBinding.privatePolicyCB.isChecked) {
                if (VMReg.isEmail(mAccount)) {
                    mViewModel.signUpByEmail(mAccount, mPassword)
                } else {
                    // TODO 只能使用邮箱注册
                    errorBar("只能使用邮箱注册账户")
                }
                // 屏蔽手机号注册
//                if (VMReg.isSimpleMobileNumber(mAccount)) {
//                    mViewModel.signUpByPhone(mAccount, mPassword)
//                }
            } else {
                errorBar(R.string.agreement_policy_hint)
            }
        }
    }

    override fun initData() {

    }

    override fun onModelLoading(model: BViewModel.UIModel) {
        mBinding.submitTV.isEnabled = !model.isLoading
        mBinding.loadingView.visibility = if (model.isLoading) View.VISIBLE else View.GONE
    }

    override fun onModelRefresh(model: BViewModel.UIModel) {
        if (model.type == "signUpByPhone" || model.type == "signUpByEmail") {
            // 登录 IM
            mViewModel.signInIM()
            // 登录成功，跳转到主页
            CRouter.goMain()
            LDEventBus.post(Constants.Event.finishPrev)
            finish()
        }
    }

    /**
     * 校验输入框内容
     */
    private fun verifyInputBox() {
        mBinding.submitTV.isEnabled = VMReg.isEmail(mAccount) && mPassword.isNotEmpty()
    }

}
